Add enable/disable unfocused functionality

omrisarig13 opened the pull request:
Added a functionality to let the plugin have an option that would ignore the marks on buffers other than the current working buffer.

It seems to me that the marks on other buffers are usually useless (at least most of them), so I added a functionality that would let the user disable showing them on buffers other then the current working buffer.

bootleq commented:
Hi, thanks for PR this feature, I have some comments:

1. Would you also add doc for `ShowMarksOff` command?
2. Please see if we can keep the doc text within 79 columns (the `*'showmarks_enable_unfocused'*` line).
3. What is the purpose of `FocusGained` and `Lost` events? From your first comment, I suppose the main problem has been solved by BufEnter/Leave events.
4. Follow above, would the option name `enable_unfocused` a bit confused to `showmarks_auto_toggle`?
  I personally feel the original `auto_toggle` not very useful, maybe extend it to support `1` `0` and `['current_buffer']` could be better.
  Or another way: not add new option, let advance users set autocmd on their own.

omrisarig13 commented:
Hey, thanks for reviewing my code.
I fixed your comments:

1. I removed this command. It was used when I started working on it, and I forgot it after I removed this function.
2. I removed the doc entirely, so there are no more lines above this limit in the file.
3. You are correct, they were useless. I removed them.
4. I fixed it as you recommended - I removed the variable of enable_unfocused and merged it with showmarks_auto_toggle. I added the new option (only for the current buffer) as 2 instead of `'current_buffer'` since it was easier (because of the if's in the code, which would stay the same with 1 and 2 but act differently for a string).

bootleq commented:
Thanks very much for the update, looks good and clean.

But I still prefer explicitly set the toggle way, Number `0` to `2` seems too simple, and currently we can't set auto toggle by "current buffer" without "CursorHold".

How about add a type check first, if is `type(0)` just go the old way.
And for `type([])`, allow user to fine tune it like `['current_buffer', 'cursor_hold']`.

I can do the implementation if you don't bother battle with this old code base.

omrisarig13 commented:
You are completely right, I haven't thought enough about the options of using the values of the cursor_hold mapping and the current_buffer mapping.

I fixed the code as you suggested. I changed the code so it would always work for a list, and if it gets a variable from type of number it convert it to the appropriate list value according to the previous values there was.

bootleq commented on plugin/showmarks.vim:
```diff
@@ -36,13 +36,25 @@ endif
+" The value of showmarks_auto_toggle was changed to be a list. For backward
+" compatibility reasons, the code still supports it as a number or as a string.
+if !exists('g:showmarks_auto_toggle' ) | let g:showmarks_auto_toggle  = ['cursor_hold'] | endif
+if !exists('v:t_number') && type(g:showmarks_auto_toggle) == type(0)
```
The `v:t_number` check seems redundant, and causes Vim 8.1 skip later logic.

omrisarig13 replied:
Yeah, it was my bad. I started by doing the condition a bit differently. When I changed the condition I forgot about this part of it.

It should work now on Vim8.1 as well.

bootleq commented:
Thank you!
I am happy to merge and prepare a new release.
